### PR TITLE
[Retiarii] Pure-python execution engine

### PR DESCRIPTION
**Pull request by ultmaster**
This PR adds an execution engine that can directly execute graph without touching any graph converter and codegen stuff. It is expected to become the default execution engine in version 2.3.

I've also implemented a more useful mutation history tracking tool. This should make the upcoming visualization thing possible.

I tried to refact/damage as little code as possible. Some of the refactoring effort might come in future releases.

~I think a lot of tests are needed before it is ready for review.~

Depending on #3607.

## User experience

Currently I don't have the confidence to make this execution engine default. Two modifications are needed to (manually) enable the new engine.

1. Add `@model_wrapper` decorator outside the full PyTorch model.
2. Add `config.execution_engine = 'py'` to `RetiariiExeConfig`.

### CODE REVIEW

**Review by liuzhe-lz — APPROVED** [review]

File: nni/retiarii/nn/pytorch/mutator.py [hunk 1]
```diff
@@ -116,12 +120,94 @@ def process_inline_mutation(model: Model) -> Optional[List[Mutator]]:
...
-
     if applied_mutators:
         return applied_mutators
     return None
 
 
+# The following are written for pure-python mode
+
+
+class ManyChooseManyMutator(Mutator):
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:131]: please add docstring for it

File: nni/retiarii/nn/pytorch/mutator.py [hunk 2]
```diff
@@ -116,12 +120,94 @@ def process_inline_mutation(model: Model) -> Optional[List[Mutator]]:
...
+            node.label = module.label
+        if isinstance(module, Placeholder):
+            raise NotImplementedError('Placeholder is not supported in python execution mode.')
+
+    model.status = ModelStatus.Frozen
+    if not graph.hidden_nodes:
+        return model, None
+
+    mutators = []
+    for nodes in _group_by_label_and_type(graph.hidden_nodes):
```
Comment (QuanluZhang) [inline nni/retiarii/nn/pytorch/mutator.py:199]: it is a little strange that both root_graph and graph are used, they are the same graph?

File: nni/retiarii/utils.py [hunk 3]
```diff
@@ -61,3 +65,35 @@ def get_module_name(cls_or_func):
 def get_importable_name(cls, relocate_module=False):
     module_name = get_module_name(cls) if relocate_module else cls.__module__
     return module_name + '.' + cls.__name__
+
+
+class ContextStack:
```
Comment (QuanluZhang) [inline nni/retiarii/utils.py:70]: better to add docstring for this class

Comment (QuanluZhang) [reply nni/retiarii/utils.py:70]: for example why we need this context manager?

File: nni/retiarii/serializer.py [hunk 4]
```diff
@@ -149,3 +151,15 @@ def basic_unit(cls):
     import torch.nn as nn
     assert issubclass(cls, nn.Module), 'When using @basic_unit, the class must be a subclass of nn.Module.'
     return serialize_cls(cls)
+
+
+def model_wrapper(cls):
```
Comment (QuanluZhang) [inline nni/retiarii/serializer.py:156]: it is not clear how to use this wrapper. what is the difference between `model_wrapper` and `serializer`?

**Review by QuanluZhang — APPROVED** [review]

**Comment by QuanluZhang** [comment]
@ultmaster looks good, but where is document for this execution engine?

File: nni/retiarii/nn/pytorch/mutator.py (hunk 1 above)
Comment (ultmaster) [reply nni/retiarii/nn/pytorch/mutator.py:131]: This is an internal API. Added a short description.

File: nni/retiarii/nn/pytorch/mutator.py (hunk 2 above)
Comment (ultmaster) [reply nni/retiarii/nn/pytorch/mutator.py:199]: I think so.

File: nni/retiarii/serializer.py (hunk 4 above)
Comment (ultmaster) [reply nni/retiarii/serializer.py:156]: I think it's documented in the docstring.

**Comment by ultmaster** [comment]
> @ultmaster looks good, but where is document for this execution engine?

I think it's written in the PR description.

The original plan is to fully test it and make this execution engine default. I didn't expect that this PR takes so longer to review. Anyway, now we are running out of time, and I think the safest plan is to document it and make this execution engine optional if user need it.

File: docs/en_US/NAS/retiarii/Advanced.rst [hunk 5]
```diff
@@ -1,7 +1,17 @@
 Advanced Tutorial
 =================
 
-This document includes two parts. The first part explains the design decision of ``@basic_unit`` and ``serializer``. The second part is the tutorial of how to write a model space with mutators.
+Pure-python execution engine (experimental)
+-------------------------------------------
+
+If you are experiencing issues with TorchScript, or the generated model code by Retiarii, there is another execution engine called Pure-python execution engine which doesn't need the code-graph conversion. This should generally not affect models and strategies in most cases, but customized mutation might not be supported.
```
Comment (duchengyao) [inline docs/en_US/NAS/retiarii/Advanced.rst:7]: 终于可以不用torchscript了，大佬牛逼

**Comment by JiaqingFu** [comment]
nice work!

**Pull request merged**